locator/tablets.hh

```diff
@@ -0,0 +1,177 @@
+
+    bool operator==(const tablet_transition_info&) const = default;
+};
+
+/// Stores information about tablets of a single table.
+class tablet_map {
+public:
+    using tablet_container = utils::chunked_vector<tablet_info>;
+private:
+    // The implementation assumes that _tablets.size() is a power of 2:
+    //
+    //   _tablets.size() == 1 << _log2_tablets
+    //
+    tablet_container _tablets;
+    size_t _log2_tablets; // log_2(_tablets.size())
+    std::unordered_map<tablet_id, tablet_transition_info> _transitions;
+public:
+    tablet_map();
+
+    /// Constructs a tablet map.
+    /// The actual number of allocated tablets may be higher due to rounding up to the power of two.
+    /// Check tablet_count().
+    ///
+    /// \param tablet_count The desired tablets to allocate.
+    explicit tablet_map(size_t tablet_count);
+
+    tablet_id get_tablet_id(token) const;
+    const tablet_info& get_tablet_info(tablet_id) const;
+    const tablet_transition_info* get_tablet_transition_info(tablet_id) const;
+    dht::token get_last_token(tablet_id id) const;
+    dht::token get_first_token(tablet_id id) const;
+    dht::token_range get_token_range(tablet_id id) const;
+
+    std::optional<tablet_id> next_tablet(tablet_id t) const {
+        if (t == tablet_count()) {
+            return std::nullopt;
+        }
+        return t + 1;
+    }
+
+    const tablet_container& tablets() const {
+        return _tablets;
+    }
+
+    size_t tablet_count() const {
+        return _tablets.size();
+    }
+
+    const tablet_info& get_tablet_info(token t) const {
+        return get_tablet_info(get_tablet_id(t));
+    }
+
+    bool operator==(const tablet_map&) const = default;
+public:
+    void set_tablet(tablet_id, tablet_info);
+    void set_tablet_transition_info(tablet_id, tablet_transition_info);
+
+    // Destroys gently.
+    // The tablet map is not usable after this call and should be destroyed.
+    future<> clear_gently();
+};
+
+/// Holds information about all tablets in the cluster.
+///
+/// When this instance is obtained via token_metadata_ptr, it is immutable
+/// (represents a snapshot) and references obtained through this are guaranteed
+/// to remain valid as long as the containing token_metadata_ptr is held.
+///
+/// Copy constructor can be invoked across shards.
+class tablet_metadata {
+public:
+    // FIXME: Make cheap to copy.
+    // We want both immutability and cheap updates, so we should use
+    // hierarchical data structure with shared pointers and copy-on-write.
+    // Currently we have immutability but updates require full copy.
+    //
+    // Also, currently the copy constructor is invoked across shards, which precludes
+    // using shared pointers. We should change that and use a foreign_ptr<> to
+    // hold immutable tablet_metadata which lives on shard 0 only.
+    // See storage_service::replica_to_all_cores().
+    using table_to_map = std::unordered_map<table_id, tablet_map>;
```

Comment:
I kinda dislike the name `table_to_map`.
How about `map_by_table` or `map_by_table_id`?

Reply:
Why do you dislike it? x_to_y is a common pattern in the code base.

Reply:
then it should be called `table_to_tablet_map`, not just `table_to_map`.
The latter is unclear about what `x` is mapped to.

Reply:
Right